### DISCUSSION TITLE
４ボタンのキーボードを作りたい❕？

### DISCUSSION THREAD

**Discussion by nakkyi** [algyan/XIAOGYAN#77] [Ideas]
XIAOGAYNの基板を使って、４キーのピアノを作ってみたいと思って、挑戦しました。
![DSC_0532](https://github.com/algyan/XIAOGYAN/assets/47250321/aa628dd1-3195-4f47-a4bd-48266c4feb24)
![DSC_0531](https://github.com/algyan/XIAOGYAN/assets/47250321/b96d2cf4-821f-4120-b692-00f6eab471af)

この基板には、Aボタン、Ｂボタンの２個しかありません。

////////////////////////////////////////////////////////////////////////////
松岡さんに相談したところ、
Xiaogyanに、Xiaogyan.buttonC、Xiaogyan.buttonDを追加することができれば分かりやすいのですが、今のライブラリの作りではアプリケーション側で追加することはできません。
buttonCとbuttonDをアプリケーション側に追加して、それを利用するのが良いと思います。

と言う事で、Grove-I2C端子を使い、Ｃボタン、Ｄボタンを追加したいと思い、次のように配線しました。
![IMG_1574](https://github.com/algyan/XIAOGYAN/assets/47250321/caa2f5cf-ac91-4706-9dd1-1996545541cf)
![IMG_1576](https://github.com/algyan/XIAOGYAN/assets/47250321/a53db950-a1cb-4f3a-8f79-8cb3508ba198)

その後、いろいろと試してみました。
// Cボタン、Dボタンで、ミ、ファを表す事が出来ました。

そして、[HardwareTest]のプログラム（Aボタン、Bボタンで「ド」「レ」「ミ」を初めて。）を参考に、次のようなプログラムを書いてみました。
’
/*
 * main.cpp
 * Copyright (C) 2023 MATSUOKA Takashi <matsujirushi@live.jp>
 * MIT License
 */

// Includes

#include <Arduino.h>
#include <elapsedMillis.h>
#include "Xiaogyan.hpp"

const int buttonC = D5;
const int buttonD = D4;

// setup and loop

void setup()
{
    Serial.begin(115200);
    delay(1000);
    Serial.println();
    Serial.println();

    // Initialize

    pinMode(buttonC, INPUT_PULLUP);
    pinMode(buttonD, INPUT_PULLUP);

    digitalWrite(buttonC, HIGH);
    digitalWrite(buttonD, HIGH);

    // XIAOGYANライブラリを初期化。
    Xiaogyan.begin();

    ////////////////////////////////////////
    // Startup Sequence

    // 「ド」音を鳴らす。
    Xiaogyan.speaker.setTone(262);  // C4
    // LEDマトリックスを全て赤色点灯。
    Xiaogyan.ledMatrix.setBrightness(2);
    Xiaogyan.ledMatrix.fillScreen(1);
    delay(200);

    // 音を止める。
    Xiaogyan.speaker.setTone(0);
    // LEDマトリックスを全て消灯。
    Xiaogyan.ledMatrix.fillScreen(0);
}

void loop()
{
    // XIAOGYANライブラリを実行。
    Xiaogyan.doWork();

    // LEDを点滅する。
    // （200ミリ秒点灯、800ミリ秒消灯を繰り返す。）
    Xiaogyan.led.write(millis() % 2000 < 200 ? LOW : HIGH);

/*
    // Cボタン、Dボタンで　ミ　ファ　を鳴らす。
    if      (digitalRead(buttonC)==LOW)  Xiaogyan.speaker.setTone(330);   // E4
    
    else if (digitalRead(buttonD)==LOW)  Xiaogyan.speaker.setTone(349);   // F4
    
    else      Xiaogyan.speaker.setTone(0);
    
    delay(40);
*/

/*
    // Aボタン、Bボタンで「ド」「レ」「ミ」を鳴らす。
    static bool buttonA = false;
    static bool buttonB = false;
    bool preButtonA = buttonA;
    bool preButtonB = buttonB;
    buttonA = Xiaogyan.buttonA.read() == LOW;
    buttonB = Xiaogyan.buttonB.read() == LOW;
    if (preButtonA != buttonA || preButtonB != buttonB)
    {
        if      ( buttonA && !buttonB) Xiaogyan.speaker.setTone(262);   // C4
        else if (!buttonA &&  buttonB) Xiaogyan.speaker.setTone(294);   // D4
        else if ( buttonA &&  buttonB) Xiaogyan.speaker.setTone(330);   // E4
        else                           Xiaogyan.speaker.setTone(0);
    }
*/
    
    // Aボタン、Bボタン、Cボタン、Dボタンで「ドレミファソラシド」を鳴らす。
    static bool buttonA = false;
    static bool buttonB = false;
    static bool buttonC = false;
    static bool buttonD = false;
    bool preButtonA = buttonA;
    bool preButtonB = buttonB;
    bool preButtonC = buttonC;
    bool preButtonD = buttonD;
    buttonA = Xiaogyan.buttonA.read() == LOW;
    buttonB = Xiaogyan.buttonB.read() == LOW;
    buttonC = digitalRead(buttonC) == LOW;
    buttonD = digitalRead(buttonD) == LOW;
    if (preButtonA != buttonA || preButtonB != buttonB || preButtonC != buttonC || preButtonD != buttonD)
    {
        if      ( buttonA && !buttonB && !buttonC && !buttonD ) Xiaogyan.speaker.setTone(262);   // C4　ド
        else if (!buttonA &&  buttonB && !buttonC && !buttonD ) Xiaogyan.speaker.setTone(294);   // D4　レ
        else if (!buttonA && !buttonB &&  buttonC && !buttonD ) Xiaogyan.speaker.setTone(330);   // E4　ミ
        else if (!buttonA && !buttonB && !buttonC &&  buttonD ) Xiaogyan.speaker.setTone(349);   // F4　ファ
        
        else if ( buttonA &&  buttonB && !buttonC && !buttonD ) Xiaogyan.speaker.setTone(392);   // G4　ソ
        else if (!buttonA &&  buttonB &&  buttonC && !buttonD ) Xiaogyan.speaker.setTone(440);   // A4　ラ
    //    else if (!buttonA && !buttonB &&  buttonC &&  buttonD ) Xiaogyan.speaker.setTone(494);   // B4　シ 
        else if ( buttonA && !buttonB && !buttonC &&  buttonD ) Xiaogyan.speaker.setTone(523);   // C5　ド
        else                           Xiaogyan.speaker.setTone(0);
    }
}
’

上記のプログラムを動かしたところ、継続継続で、どのキーを押しても構いません。ところで、下記の if 、else if の一行ずつ順番にコメントアウトしてい
ます
`
 {
        if      ( buttonA && !buttonB && !buttonC && !buttonD ) Xiaogyan.speaker.setTone(262);   // C4　ド
        else if (!buttonA &&  buttonB && !buttonC && !buttonD ) Xiaogyan.speaker.setTone(294);   // D4　レ
        else if (!buttonA && !buttonB &&  buttonC && !buttonD ) Xiaogyan.speaker.setTone(330);   // E4　ミ
        else if (!buttonA && !buttonB && !buttonC &&  buttonD ) Xiaogyan.speaker.setTone(349);   // F4　ファ
        
        else if ( buttonA &&  buttonB && !buttonC && !buttonD ) Xiaogyan.speaker.setTone(392);   // G4　ソ
        else if (!buttonA &&  buttonB &&  buttonC && !buttonD ) Xiaogyan.speaker.setTone(440);   // A4　ラ
    //    else if (!buttonA && !buttonB &&  buttonC &&  buttonD ) Xiaogyan.speaker.setTone(494);   // B4　シ (　//を取ると鳴りっぱなしになる。）
        else if ( buttonA && !buttonB && !buttonC &&  buttonD ) Xiaogyan.speaker.setTone(523);   // C5　ド
        else                           Xiaogyan.speaker.setTone(0);
    }
`
最後から5行目、

 //    else if (!buttonA && !buttonB &&  buttonC &&  buttonD ) Xiaogyan.speaker.setTone(494);   // B4　シ (　//を取ると鳴りっぱなしになる。）
を、コメントアウトしたら、音は鳴らなくなりましたが、どのボタンを押しても音が出なくなりました。

////////////////////////////////////////////////////////////////////////////
松岡さんから、下記の返事をもらいました。

間違っているのか分かりません。
対策方法をご案内します。

実機で動作して確認していませんが、、
ピン番号と読み出した値を記憶する変数名が同じなのが影響しているのかも？

...
const int buttonC = D5;
const int buttonD = D4;
...
    static bool buttonC = false;
    static bool buttonD = false;
...
    buttonC = digitalRead(buttonC) == LOW;
    buttonD = digitalRead(buttonD) == LOW;
...

///////////////////////////////////////////////////////////////////////////
そこで、次のように変更したら、作動するようになりました。

コメント、ありがとうございました。

ご指摘の通り、

...
    buttonC = digitalRead(buttonC) == LOW;
    buttonD = digitalRead(buttonD) == LOW;
...
これを、下記のように変更したら、出来るようになりました。

　buttonC = digitalRead(D5) == LOW;
    buttonD = digitalRead(D4) == LOW;
これで、ＸＩＡＯＧＹＡＮ ４キー電子ピアノの完成です。

////////////////////////////////////////////////////////////////////////////

＊　このディスカッションは、最後に完成の報告をしたのですが、
Q＆Aの #75 のやり取りを、本日のARUGYANのイベント中に消去してしまったので、完成報告も無くなってしまったので、再度書きました。(笑い)

**Comment by nakkyi** [comment]
完成動画を　YouTube　にアップしましたので、リンクを張り付けます。
https://youtu.be/jD7SCIO-KX0

**Comment by matsujirushi** [comment]
#75  に残っていますよー。
（CloseしたのでDiscussionsの一覧にデフォルトで表示されなくなっただけです。Filter: Closedを選べば一覧表示に出てきます。🙋‍♂️）

**Comment by nakkyi** [comment]
はい、了解しました。
なくなったので、消しちゃったかと思いました。
いろいろと勉強になりました。
コレからもよろしくお願いします。

2023年12月10日(日) 21:22 MATSUOKA Takashi ***@***.***>:
